libcxx/docs/Hardening.rst

```diff
@@ -72,17 +75,301 @@ to control the level by passing **one** of the following options to the compiler
+  way to completely prove that a semantic requirement is satisfied; thus, this
+  would often be a heuristic check and it might be quite expensive.
+
+- ``internal`` -- checks that internal invariants of the library hold. These
+  assertions don't depend on user input.
+
+- ``uncategorized`` -- for assertions that haven't been properly classified yet.
+  This is an escape hatch used for some existing assertions in the library; all
+  new code should have its assertions properly classified.
+
+Mapping between the hardening modes and the assertion categories
+================================================================
+
+.. list-table::
+    :header-rows: 1
+    :widths: auto
+
+    * - Category name
+      - ``fast``
+      - ``extensive``
+      - ``debug``
+    * - ``valid-element-access``
+      - ✅
+      - ✅
+      - ✅
+    * - ``valid-input-range``
+      - ✅
+      - ✅
+      - ✅
+    * - ``non-null``
+      - ❌
+      - ✅
+      - ✅
+    * - ``non-overlapping-ranges``
+      - ❌
+      - ✅
+      - ✅
+    * - ``valid-deallocation``
+      - ❌
+      - ✅
+      - ✅
+    * - ``valid-external-api-call``
+      - ❌
+      - ✅
+      - ✅
+    * - ``compatible-allocator``
+      - ❌
+      - ✅
+      - ✅
+    * - ``argument-within-domain``
+      - ❌
+      - ✅
+      - ✅
+    * - ``pedantic``
+      - ❌
+      - ✅
+      - ✅
+    * - ``semantic-requirement``
+      - ❌
+      - ❌
+      - ✅
+    * - ``internal``
+      - ❌
+      - ❌
+      - ✅
+    * - ``uncategorized``
+      - ❌
+      - ✅
+      - ✅
+
+.. note::
+
+  At the moment, each subsequent hardening mode is a strict superset of the
+  previous one (in other words, each subsequent mode only enables additional
+  assertion categories without disabling any), but this won't necessarily be
+  true for any hardening modes that might potentially be added in the future.
+
+Hardening assertion failure
+===========================
+
+In production modes (``fast`` and ``extensive``), a hardening assertion failure
+immediately traps the program. This is the safest approach that also minimizes
+the code size penalty as the failure handler maps to a single instruction. The
+downside is that the failure provides no additional details other than the stack
+trace (which might also be affected by optimizations).
+
+TODO(hardening): describe ``__builtin_verbose_trap`` once we can use it.
+
+In the ``debug`` mode, an assertion failure terminates the program in an
+unspecified manner and also outputs the associated error message to the error
+output. This is less secure and increases the size of the binary (among other
+things, to store the error message strings) but makes the failure easier to
+debug. It also allows us to test the error messages in our test suite.
+
+.. _override-assertion-handler:
+
+Overriding the assertion failure handler
+----------------------------------------
+
+Vendors can override the default termination handler mechanism by following
+these steps:
```

Comment:
Is there a way for users to override the default? If not it would be good to mention this explicitly.

Reply:
It's a good point, but I find it a little hard to address. Just stating that "users cannot override this" doesn't seem to actually provide any additional information. Stating "users _shouldn't_ override this" (i.e., we not only don't provide this mechanism but also never intend to provide it because we don't think it's technically feasible/a good idea/etc.) might be true, but I don't feel I have thought about this to say that with complete confidence. Would you be okay to leave it as is for now?

Reply:
I feel stating that "users cannot override this" does provide information. After reading that they can stop searching how they can override this. For me it makes it easier to point to the documentation when users post bug reports.

I was not aware users can't override this handle.

Reply:
Thanks for explaining, added. By the way, I just remembered it's mentioned in [P3191](https://isocpp.org/files/papers/P3191R0.pdf):
> First, under the strictest code size requirements, we do not allow users to redefine the contract violation handler. The contract violation handler is something that the vendor defines and is then baked in for all users. If the vendor wants to allow for the contract violation handler to be repleaceable, it can provide a contract violation handler that is repleaceable, but it is not required to.